File: sdk/csharp/LibraVDB/Filter.cs
using System;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace LibraVDB
{
    public class Filter
    {
        private readonly JsonObject _node;

        private Filter(JsonObject node)
        {
            _node = node;
        }

        public string ToJson()
        {
            return _node.ToJsonString();
        }

        public static Filter Eq(string field, object value) => CreateComparison("eq", field, value);
        public static Filter Gt(string field, object value) => CreateComparison("gt", field, value);
        public static Filter Lt(string field, object value) => CreateComparison("lt", field, value);
        public static Filter In(string field, object[] values) => CreateComparisonList("contains_any", field, values);

        public static Filter And(params Filter[] filters) => CreateLogical("and", filters);
        public static Filter Or(params Filter[] filters) => CreateLogical("or", filters);

        private static Filter CreateComparison(string op, string field, object value)
        {
            var node = new JsonObject
            {
                ["type"] = op,
                ["field"] = field,
                ["value"] = JsonSerializer.SerializeToNode(value)
            };
            return new Filter(node);
        }

        private static Filter CreateComparisonList(string op, string field, object[] values)
        {
            var node = new JsonObject
            {
                ["type"] = op,
                ["field"] = field,
                ["values"] = JsonSerializer.SerializeToNode(values)
            };
            return new Filter(node);
        }

        private static Filter CreateLogical(string op, Filter[] filters)
        {
            var array = new JsonArray();
            foreach (var filter in filters)
            {
                array.Add(filter._node.DeepClone());
            }

            var node = new JsonObject
            {
                ["type"] = op,
                ["filters"] = array
            };

            return new Filter(node);
        }
    }
}

File: sdk/ruby/lib/libravdb/filter.rb
module LibraVDB
  class Filter
    attr_reader :type, :args

    def initialize(type, args)
      @type = type
      @args = args
    end

    def self.eq(field, value)
      new("eq", { field: field, value: value })
    end

    def self.neq(field, value)
      new("neq", { field: field, value: value })
    end

    def self.gt(field, value)
      new("gt", { field: field, value: value })
    end

    def self.gte(field, value)
      new("gte", { field: field, value: value })
    end

    def self.lt(field, value)
      new("lt", { field: field, value: value })
    end

    def self.lte(field, value)
      new("lte", { field: field, value: value })
    end

    def self.in(field, value)
      new("in", { field: field, value: value })
    end

    def self.contains(field, value)
      new("contains", { field: field, value: value })
    end

    def self.and(*filters)
      new("and", filters.map(&:to_hash))
    end

    def self.or(*filters)
      new("or", filters.map(&:to_hash))
    end

    def self.not(filter)
      new("not", filter.to_hash)
    end

    def and(other)
      Filter.and(self, other)
    end

    def or(other)
      Filter.or(self, other)
    end

    def to_hash
      { type: @type }.merge(@args)
    end
  end
end
